File: typescript/creational/abstract-factory/IronDoorFactory.ts
import {DoorFactory} from './DoorFactory';
import {Door} from './Door';
import {IronDoor} from './IronDoor';
import {DoorFittingExpert} from './DoorFittingExpert';
import {Welder} from './Welder';

// Iron door factory to get iron door and the relevant fitting expert
export class IronDoorFactory implements DoorFactory {
  public makeDoor():Door {
    return new IronDoor();
  }

  public makeFittingExpert():DoorFittingExpert {
    return new Welder();
  }
}

File: ruby/creational/abstract-factory/IronDoorFactory.rb
require_relative 'DoorFactory'
require_relative 'IronDoor'
require_relative 'Welder'

class IronDoorFactory include DoorFactory
  def make_door
    return IronDoor.new
  end

  def make_fitting_expert
    return Welder.new
  end
end
